fortitude/src/rules/style/whitespace.rs

```diff
@@ -16,25 +17,44 @@ use crate::{FromStartEndLineCol, TextRule};
 #[violation]
 pub struct TrailingWhitespace {}
 
-impl Violation for TrailingWhitespace {
+impl AlwaysFixableViolation for TrailingWhitespace {
     #[derive_message_formats]
     fn message(&self) -> String {
         format!("trailing whitespace")
     }
+
+    fn fix_title(&self) -> String {
+        format!("Remove trailing whitespace")
+    }
 }
 impl TextRule for TrailingWhitespace {
-    fn check(_settings: &Settings, source: &SourceFile) -> Vec<Diagnostic> {
+    fn check(_settings: &Settings, source_file: &SourceFile) -> Vec<Diagnostic> {
+        let source = source_file.to_source_code();
         let mut violations = Vec::new();
-        for (idx, line) in source.source_text().split('\n').enumerate() {
-            if line.ends_with([' ', '\t']) {
-                violations.push(Diagnostic::from_start_end_line_col(
-                    Self {},
-                    source,
-                    idx,
-                    line.trim_end().len(),
-                    idx,
-                    line.len(),
+        for (idx, line) in source.text().lines().enumerate() {
+            let whitespace_bytes: TextSize = line
+                .chars()
+                .rev()
+                .take_while(|c| c.is_whitespace())
+                .map(TextLen::text_len)
+                .sum();
+            if whitespace_bytes > 0.into() {
+                let line_end_byte = source.line_end_exclusive(OneIndexed::from_zero_indexed(idx));
+                let edit = Edit::range_deletion(TextRange::new(
+                    line_end_byte - whitespace_bytes,
+                    line_end_byte,
                 ));
+                violations.push(
+                    Diagnostic::from_start_end_line_col(
+                        Self {},
+                        source_file,
+                        idx,
+                        line.trim_end().len(),
+                        idx,
+                        line.len(),
```

Comment:
Should we now have:

```suggestion
                        line_end_byte - whitespace_bytes,
                        idx,
                        line_end_byte,
```

Reply:
No, because `line_end_byte` is a bytes offset into the whole file, while `Diagnostic::from_start_end_line_col` uses zero-indexed rows and cols. However, what we can do is just use Ruff's `Diagnostic::new(kind, text_range)` directly, as we already have the `TextRange` here.

Reply:
Ah, nice!